src/desktop/credentials/DesktopNativeCredentialsFacade.ts

```diff
@@ -1,35 +1,136 @@
-		this._desktopKeyStoreFacade = keyStoreFacade
-		this._crypto = crypto
+	async decryptUsingKeychain(data: Uint8Array, encryptionMode: DesktopCredentialsMode): Promise<Uint8Array> {
+		// making extra sure that the mode is the right one since this comes over IPC
+		this.assertSupportedEncryptionMode(encryptionMode)
+		data = await this.maybeDecryptUsingAppPin(data, encryptionMode)
+		const credentialsKey = await this.desktopKeyStoreFacade.getCredentialsKey()
+		data = this.crypto.aes256DecryptKey(credentialsKey, data)
+		return data
 	}
 
-	async decryptUsingKeychain(data: Uint8Array, encryptionMode: CredentialEncryptionMode.DEVICE_LOCK): Promise<Uint8Array> {
+	async encryptUsingKeychain(data: Uint8Array, encryptionMode: DesktopCredentialsMode): Promise<Uint8Array> {
 		// making extra sure that the mode is the right one since this comes over IPC
-		assert(encryptionMode === CredentialEncryptionMode.DEVICE_LOCK, "should not use unsupported encryption mode")
-		const key = await this._desktopKeyStoreFacade.getCredentialsKey()
-		const decryptedData = this._crypto.aes256DecryptKey(key, data)
-		return Promise.resolve(decryptedData)
+		this.assertSupportedEncryptionMode(encryptionMode)
+		const credentialsKey = await this.desktopKeyStoreFacade.getCredentialsKey()
+		data = this.crypto.aes256EncryptKey(credentialsKey, data)
+		data = await this.maybeEncryptUsingAppPin(data, encryptionMode)
+		return data
 	}
 
-	async encryptUsingKeychain(data: Uint8Array, encryptionMode: CredentialEncryptionMode.DEVICE_LOCK): Promise<Uint8Array> {
-		// making extra sure that the mode is the right one since this comes over IPC
-		assert(encryptionMode === CredentialEncryptionMode.DEVICE_LOCK, "should not use unsupported encryption mode")
-		const key = await this._desktopKeyStoreFacade.getCredentialsKey()
-		const encryptedData = this._crypto.aes256EncryptKey(key, data)
-		return Promise.resolve(encryptedData)
+	private async maybeDecryptUsingAppPin(data: Uint8Array, encryptionMode: DesktopCredentialsMode): Promise<Uint8Array> {
+		if (encryptionMode === CredentialEncryptionMode.APP_PIN) {
+			const appPinKey = await this.resolveKeyFromAppPin(false)
+			if (appPinKey == null) {
+				// there is no salt, so someone might have turned off app pin in the meantime.
+				// if we actually have pin-encrypted credentials and just lost the salt, we
+				// will get a decryption failure and delete the credentials further up the stack.
+				return data
+			}
+			try {
+				data = this.crypto.aesDecryptBytes(appPinKey, data)
+			} catch (e) {
+				if (e instanceof CryptoError) {
+					const nativeFacade = await this.getCurrentCommonNativeFacade()
+					nativeFacade.showAlertDialog("invalidPassword_msg")
+					throw new CancelledError("app pin verification failed")
+				} else {
+					throw e
+				}
+			}
+			return data
+		} else {
+			// our mode is not app pin, so the app pin should not be set.
```

Comment:
The only reason we need it is because we always return a key from `resolveKeyFromAppPin` (see my comment there). If we change that one we wouldn't need this or this whole method.